core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -821,6 +827,12 @@ private Builder(TableMetadata base) {
       this.sortOrdersById = Maps.newHashMap(base.sortOrdersById);
     }
 
+    public Builder withMetadataLocation(String newMetadataLocation) {
+      this.metadataLocation = newMetadataLocation;
+      this.discardChanges = true;
```

Comment:
What's the motivation for discarding changes automatically in this case?  Is it that we don't want to apply on top of unknown metadata?  A comment here would be helpful since we're performing a hidden action on this builder call.

Reply:
I'll fix this. I think I want the caller to specifically call `discardChanges()` or fail validation if there are any changes.

The contract for `TableMetadata` is that the contents of the metadata location matches the table metadata. That's because the only way to get a `TableMetadata` with a non-null metadata location was to read it from the location using `TableMetadataParser`.

Now, table metadata is coming from a REST endpoint that also sends `metadataLocation`, so we need to be able to read the table metadata and re-attach the metadata location.

Reply:
Added a precondition to the build method and removed this.